Feature/1223 order badge take two

nbradbury opened the pull request:
Closes #1223 - replaces #1228, which got too ambitious with its changes. This PR simply adds a badge to the order list to indicate the unfilled order count.

@kyleaparker I also updated the badge to more closely match the design in Figma. If this will suffice, then there's no need for an SVG.

![Screenshot_1563041952](https://user-images.githubusercontent.com/3903757/61175257-06c9ea80-a57a-11e9-8273-1c554147efe9.png)

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

loremattei commented:
I'm moving this to 2.4 because 2.3 has been cut. If you want this to make it to 2.3, please feel free to move it back, target the release branch and ping me to build a new beta.

kyleaparker commented:
This looks good! Can you add a shot of the badge with a single number? The badge should be minimum 13px x 13px and I have the border radius set to 13px

nbradbury commented:
> This looks good! Can you add a shot of the badge with a single number?

@kyleaparker Here you go!

![Screenshot_1563204892](https://user-images.githubusercontent.com/3903757/61228666-bb881700-a6f4-11e9-913d-421015a0936b.png)

kyleaparker commented:
Looks good!

nbradbury commented:
@AmandaRiu Just wanted to give you a heads-up that I assigned this to you since you looked at the previous PR.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/notifications/NotifsListAdapter.kt:
```diff
@@ -230,13 +217,13 @@ class NotifsListAdapter @Inject constructor(context: Context) : SectionedRecycle
             }
             if (currentPos == position) {
-                return if (notif.read) ItemType.READ_NOTIF else ItemType.UNREAD_NOTIF
+                return if (notif.read) ItemType.UNBADGED else ItemType.BADGED
```
I wanted to name this in a way that wasn't specific to notifications or unread state because I suspect we'll reuse this item decoration for the order list. I'm happy to use terminology other than badging but chose that because it's tied to an actual badge (ie: the badge on the notifs list reflects unread notifs, and we use this item decoration on unread notifs).

nbradbury replied:
Ignore that ⬆️. I'll go ahead and change it back to read/unread, and we reuse this elsewhere we can rename it again.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt:
```diff
@@ -132,6 +147,39 @@ class MainPresenter @Inject constructor(
+
+                // TODO: May want to change the name of this stat
+                AnalyticsTracker.track(
+                        Stat.DASHBOARD_UNFULFILLED_ORDERS_LOADED,
```
I think we might want to wait to rename this until the feature is added on iOS. WDYT?

AmandaRiu replied:
IMO we should do it now so it's clear this event changed when the view changed, but I'd like to hear why you'd want to wait - there may be something I haven't considered. I would hold off on deprecating the old event until it's no longer being fired for sure.

nbradbury replied:
> I would hold off on deprecating the old event until it's no longer being fired for sure.

I'm good with that. I'll change the event name.

nbradbury replied:
Done in 10af22b

AmandaRiu replied:
🙇‍♀

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/notifications/NotifsListFragment.kt:
```diff
@@ -138,8 +138,7 @@ class NotifsListFragment : TopLevelFragment(),
 
-        val unreadDecoration = NotifsListItemDecoration(activity as Context)
-        unreadDecoration.setListener(this)
+        val unreadDecoration = BadgedItemDecoration(activity as Context, this)
```
I'll go ahead and rename this as you suggest. If it turns out we reuse this widget elsewhere we can rename it to something else.

AmandaRiu replied:
@nbradbury I thought `UnreadListItemDecoration` would be generic enough to make it reusable for the future without name changes. WDYT?

nbradbury commented:
> * The orders unfilled count badge does not update after the order list updates.

Good catch. I'll change it so the unfilled count is checked whenever the order list is updated.

> * For the same reasons as above, fulfilling an order does not change the badge count.

I'm not seeing this. There's code [here](https://github.com/woocommerce/woocommerce-android/blob/f400fe736a145f8af2b4f3020fe73f0489350f2d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt#L175) which detects when an order status has changed and updates the count. Perhaps the issue is that there's a delay between the status changing and the badge reflecting the new count?

> * Orders badge count does not update when a new order notification is received and the app is already open (we should also refresh the orders list in this case now that orders include notifications)

Another good catch. I'll correct this.

AmandaRiu commented:
> I'm not seeing this. There's code here which detects when an order status has changed and updates the count. Perhaps the issue is that there's a delay between the status changing and the badge reflecting the new count?

Yeah I was having a bunch of network issues last night so I wouldn't be surprised if it was a delay. I wonder if maybe we should do two things:
- Fetch an updated count when the order list is refreshed - which you already mentioned. This will ensure the values are as accurate as the list we current have loaded
- Add a method to manually increment/decrement the count so it can be forced to update whenever a change happens successfully while we wait for the network to get us the most updated count.

nbradbury commented:
@AmandaRiu Ready for another round!

nbradbury commented:
> * If you mark an order complete, the badge count decrements immediately, but if you UNDO the fulfillment request, the badge count doesn't revert back to the proper count.

🤦‍♂ Of course, I should've realized this. I've just addressed this, but I wasn't comfortable having the count reverted from the order list. Instead, when the order's fulfilled status changes in the order list, it tells the main activity to fetch the order count again and hide the count (but not the badge) until the request completes.

> * one of my nitpick comments was not addressed,

Ah, I missed the one about using "badge" to describe the unread item decoration. Fixed.
